index/scorch/persister.go

```diff
@@ -136,15 +141,55 @@ OUTER:
 		case <-w.notifyCh:
 			// woken up, next loop should pick up work
 			continue OUTER
-		case ew := <-s.persisterNotifier:
+		case ew = <-s.persisterNotifier:
 			// if the watchers are already caught up then let them wait,
 			// else let them continue to do the catch up
 			persistWatchers = append(persistWatchers, ew)
-			notifyWatchers()
 		}
 	}
 }
 
+func notifyMergeWatchers(lastPersistedEpoch uint64,
+	persistWatchers []*epochWatcher) []*epochWatcher {
+	var watchersNext []*epochWatcher
+	for _, w := range persistWatchers {
+		if w.epoch < lastPersistedEpoch {
+			close(w.notifyCh)
+		} else {
+			watchersNext = append(watchersNext, w)
+		}
+	}
+	return watchersNext
+}
+
+func (s *Scorch) pausePersisterForMergerCatchUp(lastPersistedEpoch uint64, lastMergedEpoch *uint64,
+	persistWatchers []*epochWatcher) []*epochWatcher {
+
+OUTER:
+	for {
+
+		// first, let the watchers proceed if they lag behind
+		persistWatchers = notifyMergeWatchers(lastPersistedEpoch, persistWatchers)
+
+		// check for slow merger and pause persister until merger catch up
+		if lastPersistedEpoch > *lastMergedEpoch+epochDistance {
+
+			select {
+			case <-s.closeCh:
+				break OUTER
+			case ew := <-s.persisterNotifier:
+				persistWatchers = append(persistWatchers, ew)
+				*lastMergedEpoch = ew.epoch
+				continue OUTER
+			}
+		} else {
```

Comment:
one thought here is that we have an unconditional for loop, but this else block is a case you check every loop iteration to bail on the loop.  so, wouldn't it be more clear to invert that (since it's an else case) and make that the condition of the for loop?

the only downside i see is that right now you always call notifyMergeWatchers regardless of this condition, you'd need it before the for loop, and again inside the loop (presumably now at the end of the loop).  but, even though that makes it feel longer, you can drop the else case, and the "continue OUTER" becomes unnecessary too as you'd want to proceed to notifyMergeWatchers again and let the loop naturally continue without need a continue